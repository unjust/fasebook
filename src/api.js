import axios from 'axios';
import jwt from 'jsonwebtoken';

const accessTokenSecret = process.env.REACT_APP_TOKEN_SECRET;

export const authenticateUser = function(username, password) {
  // send username and password to db
  // wait for response with jwt or not
  // redirect
  debugger
  axios.post('/api/auth', { username, password})
    .then((resp) => {
      console.log(resp);
      sessionStorage.setItem('fasebookToken', resp.data);
    })
    .catch((e) => {
      console.log(e);
    })
}

export const validateUser = function() {
  const t = sessionStorage.getItem('fasebookToken');
  try {
    jwt.verify(t, accessTokenSecret);
    return true;
  } catch(err) {
    console.log('error validating token', err);
    return false;
  }
}
