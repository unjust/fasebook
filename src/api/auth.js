import axios from 'axios';
import jwt from 'jsonwebtoken';

export const accessTokenSecret = process.env.REACT_APP_TOKEN_SECRET;
export const tokenSessionStorageKey = 'fasebookToken';

export const getAuthToken = () => sessionStorage.getItem(tokenSessionStorageKey);

// organized this way for testing purposes
export const authData = {
  store: (userId, accessToken) => {
    sessionStorage.setItem('userId', userId);
    sessionStorage.setItem(tokenSessionStorageKey, accessToken); 
    console.log(sessionStorage.getItem(tokenSessionStorageKey));  
  }
};

export const validateUserToken = function() {
  const t = getAuthToken();
  try {
    jwt.verify(t, accessTokenSecret);
    return true;
  } catch(err) {
    console.log('token not validated', err);
    return false;
  }
}

export const authenticateUser = (username, password) =>
  axios.post('/api/auth', { username, password })
    .then((resp) => {
      // TODO move this to catch?
      if (resp.data && resp.data.error) {
        return 'Usario o clave estan incorrectos';
      }
      const { userId, accessToken } = resp.data;
      authData.store(userId, accessToken);
      // TODO need to figure out responses for unsuccessful logins
      return;
    })
    .catch((e) => {
      console.log(e);
    })



