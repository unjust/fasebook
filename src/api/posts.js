import axios from 'axios';
import { getAuthToken } from './auth';

let axiosInstance;

const instance = () => {
  if (axiosInstance) {
    return axiosInstance;
  }
  axiosInstance = axios.create({
    headers: { Authorization: `Bearer ${getAuthToken()}` }
  });
  return axiosInstance;
};

export const getPosts = (userId) =>
  instance().get(`/api/posts/${userId}`)
    .then((resp) => resp.data)
    .catch((resp) => resp.error);

export const savePost = (userId, postContent) =>
  instance().post('/api/post', { post: postContent, userId })
    .then((resp) => resp.data)
    .catch((resp) => resp.error);

export const editPost = (userId, postId, postContent) =>
  instance().put('/api/post', { post: postContent, userId, postId })
    .then((resp) => resp.data)
    .catch((resp) => resp.error);

export const deletePost = (userId, postId) =>
  instance().delete(`/api/post/${postId}/${userId}`)
    .then((resp) => resp.data)
    .catch((resp) => resp.error);
