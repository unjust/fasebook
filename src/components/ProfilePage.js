import React from 'react';
import PropTypes from 'prop-types';
import * as postsApi from '../api/posts';
import Post from './Post';
import '../scss/profilePage.scss';

export default class ProfilePage extends React.Component {
  constructor(props) {
    super(props);
    this.postTextInput = React.createRef();
    this.handleSavePost = this.handleSavePost.bind(this);
    this.handleEditPost = this.handleEditPost.bind(this);
    this.handleDeletePost = this.handleDeletePost.bind(this);
    this.state = { posts: [] };
  }

  componentDidMount() {
    // call api for current posts
    postsApi.getPosts(this.props.userId)
      .then((posts) => {
        this.setState({ posts });
      })
      .catch(() => {});
  }

  handleSavePost() {
    postsApi.savePost(this.props.userId, this.postTextInput.current.value)
      .then((savedPost) => {
        this.postTextInput.current.value = '';
        const posts = [savedPost, ...this.state.posts];
        this.setState({ posts });
      });
  }

  handleEditPost(postId, newText) {
    postsApi.editPost(this.props.userId, postId, newText)
      // eslint-disable-next-line no-console
      .then(() => console.log('didnt implement'))
      .catch(() => {
        // eslint-disable-next-line no-alert
        alert('an error ocurred');
      });
  }

  handleDeletePost(postId) {
    postsApi.deletePost(this.props.userId, postId)
      .then(() => {
        const posts = this.state.posts.filter((p) => p.id !== postId);
        this.setState({ posts });
      });
  }

  renderPosts() {
    const nodes = this.state.posts.reduce((nodeArr, p) => {
      nodeArr.push(
        <Post
          {...p}
          key={p.id}
          isEditable={this.props.isOwner || undefined}
          onEdit={this.handleEditPost}
          onDelete={this.handleDeletePost} />
      );
      return nodeArr;
    }, []);
    return <div>{nodes}</div>;
  }

  render() {
    return (
      <div>
        <div className='card'>
          <div id='profile-head' className='flex'>
            <div className='flex-item--shrink'>
              <img className='img--thumbnail'
              src="https://scontent-bos3-1.xx.fbcdn.net/v/t1.0-9/39992261_1780642188639098_3104100502657302528_n.jpg?_nc_cat=103&_nc_sid=85a577&_nc_ohc=PeMsRB9nnxwAX_kIRHZ&_nc_ht=scontent-bos3-1.xx&oh=898a95c20b993199c04c7a53325f3f5c&oe=5ED4FF45"
              alt='your fase' />
            </div>
            <h1 className='flex-item'>CRISTIAN CASTRO</h1>
          </div>
        </div>
      { this.props.isOwner &&
        <div className='card flex--column'>
            <textarea
              ref={this.postTextInput}
              name='postContent'
              placeholder='Que quieres compartir?'
              className='margin--bottom'
              onChange={(e) => this.setState({ enablePost: !!e.target.value.trim() })}></textarea>
            <button
              className='button align--right'
              disabled={!this.state.enablePost}
              onClick={this.handleSavePost}>POST</button>
          </div>
        }
        {this.renderPosts()}
      </div>
    );
  }
}

ProfilePage.propTypes = {
  userId: PropTypes.string,
  isOwner: PropTypes.bool
};
