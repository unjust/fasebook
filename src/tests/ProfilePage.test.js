import React from 'react';
import TestRenderer from 'react-test-renderer';
import ProfilePage from '../components/ProfilePage';
import * as postsApi from '../api/posts';


describe('ProfilePage testRenderer', () => {

  jest.mock('../api/posts');
  const props = {
    userId: 'axifg345',
    isOwner: true
  };

  let testRenderer;

  it('renders the testRenderer that matches the snapshot', () => {
    testRenderer = TestRenderer.create(
      <ProfilePage {...props} />
    );
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('getPosts and render', () => {
    it('calls getPosts on mount', (done) => {
      postsApi.getPosts = jest.fn(() =>
        new Promise((resolve) => {
          process.nextTick(() => resolve([]));
        }));
      TestRenderer.create(
        <ProfilePage {...props} />
      );
      expect(postsApi.getPosts).toHaveBeenCalledWith(props.userId);
      done();
    });

    it('calls renderPosts', (done) => {
      const renderPostsSpy = jest.spyOn(ProfilePage.prototype, 'renderPosts');
      TestRenderer.create(<ProfilePage {...props} />);
      expect(renderPostsSpy).toHaveBeenCalled();
      done();
    });
  });

  describe('handlers call appropriate apis', () => {
    it('calls savePost on handleSavePost', (done) => {
      const inputValue = 'blahblah';
      const savePostSpy = jest.spyOn(postsApi, 'savePost');

      testRenderer = TestRenderer.create(
        <ProfilePage {...props} />,
        {
          createNodeMock: (element) => {
            if (element.type === 'textarea') {
              return { value: inputValue };
            }
            return null;
          }
        }
      );

      testRenderer.root.instance.handleSavePost();
      expect(savePostSpy).toHaveBeenCalledWith(props.userId, inputValue);
      done();
    });

    it('calls editPost on handleEditPost', () => {
      const postId = '100001';
      const text = 'test text';
      const editPostSpy = jest.spyOn(postsApi, 'editPost').mockImplementation(() => Promise.resolve());
      testRenderer = TestRenderer.create(<ProfilePage {...props} />);
      testRenderer.root.instance.handleEditPost(postId, text);
      expect(editPostSpy).toHaveBeenCalledWith(props.userId, postId, text);
    });

    it('calls deletePost on handleDeletePost', () => {
      const postId = '100001';
      const deletePostSpy = jest.spyOn(postsApi, 'deletePost').mockImplementation(() => Promise.resolve());
      testRenderer = TestRenderer.create(<ProfilePage {...props} />);
      testRenderer.root.instance.handleDeletePost(postId);
      expect(deletePostSpy).toHaveBeenCalledWith(props.userId, postId);
    });
  });
});
