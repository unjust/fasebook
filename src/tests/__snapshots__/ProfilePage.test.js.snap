// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage component renders the component that matches the snapshot 1`] = `
<div>
  <div
    className="card"
  >
    <div
      className="flex"
      id="profile-head"
    >
      <div
        className="flex-item--shrink"
      >
        <img
          alt="your fase"
          className="img--thumbnail"
          src="https://scontent-bos3-1.xx.fbcdn.net/v/t1.0-9/39992261_1780642188639098_3104100502657302528_n.jpg?_nc_cat=103&_nc_sid=85a577&_nc_ohc=PeMsRB9nnxwAX_kIRHZ&_nc_ht=scontent-bos3-1.xx&oh=898a95c20b993199c04c7a53325f3f5c&oe=5ED4FF45"
        />
      </div>
      <h1
        className="flex-item"
      >
        CRISTIAN CASTRO
      </h1>
    </div>
  </div>
  <div
    className="card flex--column"
  >
    <textarea
      className="margin--bottom"
      name="postContent"
      onChange={[Function]}
      placeholder="Que quieres compartir?"
    />
    <button
      className="button align--right"
      disabled={true}
      onClick={[Function]}
    >
      POST
    </button>
  </div>
  <div />
</div>
`;

exports[`ProfilePage testRenderer renders the testRenderer that matches the snapshot 1`] = `
<div>
  <div
    className="card"
  >
    <div
      className="flex"
      id="profile-head"
    >
      <div
        className="flex-item--shrink"
      >
        <img
          alt="your fase"
          className="img--thumbnail"
          src="https://scontent-bos3-1.xx.fbcdn.net/v/t1.0-9/39992261_1780642188639098_3104100502657302528_n.jpg?_nc_cat=103&_nc_sid=85a577&_nc_ohc=PeMsRB9nnxwAX_kIRHZ&_nc_ht=scontent-bos3-1.xx&oh=898a95c20b993199c04c7a53325f3f5c&oe=5ED4FF45"
        />
      </div>
      <h1
        className="flex-item"
      >
        CRISTIAN CASTRO
      </h1>
    </div>
  </div>
  <div
    className="card flex--column"
  >
    <textarea
      className="margin--bottom"
      name="postContent"
      onChange={[Function]}
      placeholder="Que quieres compartir?"
    />
    <button
      className="button align--right"
      disabled={true}
      onClick={[Function]}
    >
      POST
    </button>
  </div>
  <div />
</div>
`;
